File: sdk/python/src/qaraa/stream.py
"""Reconnectable, cancellation-safe snapshot iterator."""
# Licensed under the Apache License, Version 2.0.
from __future__ import annotations
from urllib.parse import quote, urlencode
from .codec import decode_event
from .errors import QaraaError, TransportError, exception_from_envelope
from .models import ErrorEnvelope, SnapshotUpdatedEvent

class QaraaSnapshotStream:
    def __init__(self, client, session_id: str, revision: int):
        if isinstance(revision,bool) or not isinstance(revision,int) or revision<0 or revision>9007199254740991: raise ValueError("last_snapshot_revision must be a non-negative safe integer")
        self._client,self._session,self._revision=client,session_id,revision
        self._closed,self._socket=False,None
        self._generation=0
    def __aiter__(self): return self
    async def __anext__(self):
        delays=(.1,.25,.5,1,2); failures=0
        while not self._closed and not self._client._closed:
            if self._socket is None:
                generation=self._generation
                query=urlencode({"protocolVersion":1,"lastSnapshotRevision":self._revision,"requestId":self._client._request_id()})
                base=self._client._base_url.replace("https://","wss://",1).replace("http://","ws://",1)
                url=f"{base}/v1/sessions/{quote(self._session,safe='')}/stream?{query}"
                try:
                    connector=self._client._connect
                    if connector is None:
                        from websockets.asyncio.client import connect
                        connector=connect
                    candidate=await connector(url,max_size=self._client._limit)
                    if self._closed or self._client._closed or generation!=self._generation:
                        try: await candidate.close()
                        except Exception: pass
                        raise StopAsyncIteration
                    self._socket=candidate
                except StopAsyncIteration: raise
                except Exception as error:
                    if failures>=len(delays): raise TransportError("QARAA WebSocket reconnect limit exhausted") from error
                    await self._client._delay(delays[failures]); failures+=1; continue
            try:
                async for frame in self._socket:
                    if self._closed or self._client._closed: raise StopAsyncIteration
                    if isinstance(frame,(bytes,str)) and len(frame if isinstance(frame,bytes) else frame.encode())>self._client._limit:
                        raise TransportError("QARAA WebSocket message exceeds configured size limit")
                    event=decode_event(frame,max_bytes=self._client._limit)
                    if isinstance(event,ErrorEnvelope): raise exception_from_envelope(event)
                    if isinstance(event,SnapshotUpdatedEvent) and event.session_id==self._session and event.snapshot.revision>self._revision:
                        self._revision=event.snapshot.revision; return event.snapshot
                await self._socket.close(); self._socket=None
                if failures>=len(delays): raise TransportError("QARAA WebSocket reconnect limit exhausted")
                await self._client._delay(delays[failures]); failures+=1
            except StopAsyncIteration: raise
            except (QaraaError, TransportError): raise
            except Exception:
                if self._closed or self._client._closed: raise StopAsyncIteration
                socket,self._socket=self._socket,None
                try: await socket.close()
                except Exception: pass
                if failures>=len(delays): raise
                await self._client._delay(delays[failures]); failures+=1
        raise StopAsyncIteration
    async def aclose(self):
        if self._closed:return
        self._closed=True
        self._generation+=1
        socket,self._socket=self._socket,None
        if socket is not None:
            try: await socket.close()
            except Exception: pass

File: sdk/go/qaraa/stream.go
// Licensed under the Apache License, Version 2.0.
package qaraa

import (
	"context"
	"errors"
	"github.com/coder/websocket"
	"net/url"
	"strings"
)

func (c *Client) Stream(ctx context.Context, sessionID string, lastRevision Revision) (<-chan ReadingSnapshot, <-chan error) {
	snapshots := make(chan ReadingSnapshot, 1)
	failures := make(chan error, 1)
	if lastRevision > maxSafeInteger {
		close(snapshots)
		failures <- errors.New("last revision exceeds JSON safe integer range")
		close(failures)
		return snapshots, failures
	}
	streamCtx, cancel := context.WithCancel(ctx)
	c.mu.Lock()
	c.streamSeq++
	id := c.streamSeq
	if c.closed.Load() {
		cancel()
	}
	c.streams[id] = cancel
	c.mu.Unlock()
	go func() {
		defer close(snapshots)
		defer close(failures)
		defer cancel()
		defer func() { c.mu.Lock(); delete(c.streams, id); c.mu.Unlock() }()
		revision := lastRevision
		attempt := 0
		for {
			if c.closed.Load() {
				return
			}
			select {
			case <-streamCtx.Done():
				return
			default:
			}
			target, err := c.endpoint(sessionID, "/stream")
			if err != nil {
				failures <- err
				return
			}
			target = strings.Replace(target, "https://", "wss://", 1)
			target = strings.Replace(target, "http://", "ws://", 1)
			query := url.Values{"protocolVersion": {"1"}, "lastSnapshotRevision": {revisionString(revision)}, "requestId": {c.requestID()}}
			target += "?" + query.Encode()
			connection, response, err := c.dial(streamCtx, target, &websocket.DialOptions{HTTPHeader: c.headers})
			if response != nil && response.Body != nil {
				response.Body.Close()
			}
			if err != nil {
				if !c.streamRetry(streamCtx, &attempt) {
					if streamCtx.Err() == nil {
						failures <- err
					}
					return
				}
				continue
			}
			connection.SetReadLimit(c.maxBytes)
			for {
				_, data, readErr := connection.Read(streamCtx)
				if readErr != nil {
					_ = connection.CloseNow()
					if !c.streamRetry(streamCtx, &attempt) {
						if streamCtx.Err() == nil {
							failures <- readErr
						}
						return
					}
					break
				}
				event, decodeErr := DecodeEventLimit(data, c.maxBytes)
				if decodeErr != nil {
					_ = connection.Close(websocket.StatusUnsupportedData, "invalid QARAA event")
					failures <- decodeErr
					return
				}
				if protocolErr, ok := event.(*Error); ok {
					_ = connection.Close(websocket.StatusPolicyViolation, string(protocolErr.Code))
					failures <- protocolErr
					return
				}
				updated, ok := event.(SnapshotUpdatedEvent)
				if !ok || updated.SessionID != sessionID || updated.Snapshot.Revision <= revision {
					continue
				}
				revision = updated.Snapshot.Revision
				attempt = 0
				select {
				case snapshots <- updated.Snapshot:
				default:
					select {
					case <-snapshots:
					default:
					}
					select {
					case snapshots <- updated.Snapshot:
					case <-streamCtx.Done():
						return
					}
				}
			}
		}
	}()
	return snapshots, failures
}
func revisionString(v Revision) string {
	if v == 0 {
		return "0"
	}
	digits := make([]byte, 0, 20)
	for v > 0 {
		digits = append(digits, byte('0'+v%10))
		v /= 10
	}
	for i, j := 0, len(digits)-1; i < j; i, j = i+1, j-1 {
		digits[i], digits[j] = digits[j], digits[i]
	}
	return string(digits)
}
func (c *Client) streamRetry(ctx context.Context, attempt *int) bool {
	if *attempt >= c.retry.MaxAttempts-1 {
		return false
	}
	*attempt++
	return c.delay(ctx, *attempt) == nil && !c.closed.Load()
}

var _ = errors.Is
